File: design_patterns/structural/proxy/go/internal/user/proxy.go
package user

import (
	"log"
	"sync"
)

type ServiceProxy struct {
	realService Service
	cache       map[int]string
	mu          sync.RWMutex
}

func NewServiceProxy(realService Service) *ServiceProxy {
	return &ServiceProxy{
		realService: realService,
		cache:       make(map[int]string),
	}
}

func (p *ServiceProxy) GetUser(id int) (string, error) {
	p.mu.RLock()
	user, found := p.cache[id]
	p.mu.RUnlock()

	if found {
		log.Printf("Cache hit for the user %d\n", id)
		return user, nil
	}

	log.Printf("Cache miss, calling real service for user %d\n", id)
	user, err := p.realService.GetUser(id)
	if err != nil {
		return "", err
	}

	p.mu.Lock()
	p.cache[id] = user
	p.mu.Unlock()
	return user, nil
}

File: design_patterns/structural/proxy/python/src/proxy/user/proxy.py
import threading

from .service import Service


class ProxyService(Service):
    def __init__(self, realService: Service) -> None:
        self._realservice = realService
        self._cache = {}
        self._lock = threading.Lock()

    def GetUser(self, id: int) -> str:
        with self._lock:
            if id in self._cache:
                print(f"Cache hit for the user {id}")
                return self._cache[id]

        print(f"Cache miss, calling real service for user {id}")
        user = self._realservice.GetUser(id)

        with self._lock:
            self._cache[id] = user

        return user
